### PR TITLE
add symbol clip

### PR DESCRIPTION
**Pull request by yajiedesign**
add symbol clip
remove old clip

### CODE REVIEW

File: src/engine/profiler.cc [hunk 1]
```diff
@@ -26,6 +26,7 @@ Profiler::Profiler()
   // TODO(ziheng) get device number during execution
   int kMaxNumCpus = 64;
   this->cpu_num_ = kMaxNumCpus;
```
Comment (howard0su) [inline src/engine/profiler.cc]: better you submit a separate PR.

**Comment by sxjscience** [comment]
Fix the test and we can merge.

File: src/operator/tensor/matrix_op-inl.h [hunk 2]
```diff
@@ -704,6 +715,59 @@ inline bool CropShape(const nnvm::NodeAttrs& attrs,
...
+  CHECK_EQ(inputs[0].type_flag_, outputs[0].type_flag_);
+  Stream<xpu> *s = ctx.get_stream<xpu>();
+
+  MSHADOW_TYPE_SWITCH(outputs[0].type_flag_, DType, {
+    Tensor<xpu, 1, DType> in = inputs[0].FlatTo1D<xpu, DType>(s);
+    Tensor<xpu, 1, DType> out = outputs[0].FlatTo1D<xpu, DType>(s);
+
+    out = F<mshadow_op::ClipMax::mshadow_op>(
+      F<mshadow_op::ClipMin::mshadow_op>(in, scalar(DType(param.a_min))),
+      scalar(DType(param.a_max)));
```
Comment (howard0su) [inline src/operator/tensor/matrix_op-inl.h]: can we support when a_max = inf, and a_min = -inf? so we can skip one step

File: src/engine/profiler.cc (hunk 1 above)
Comment (yajiedesign) [inline src/engine/profiler.cc]: ok https://github.com/dmlc/mxnet/pull/4747/files

File: python/mxnet/optimizer.py [hunk 3]
```diff
@@ -622,7 +622,7 @@ def update(self, index, weight, grad, state):
 
         grad = grad * self.rescale_grad
         if self.clip_gradient is not None:
-            grad = clip(grad, -self.clip_gradient, self.clip_gradient)
+            grad = clip(grad, a_min=-self.clip_gradient, a_max=self.clip_gradient)
```
Comment (piiswrong) [inline python/mxnet/optimizer.py]: Is this necessary? It shouldn't be.

File: src/operator/tensor/matrix_op.cc [hunk 4]
```diff
@@ -13,6 +13,7 @@ DMLC_REGISTER_PARAMETER(ReshapeParam);
 DMLC_REGISTER_PARAMETER(TransposeParam);
 DMLC_REGISTER_PARAMETER(ExpandDimParam);
 DMLC_REGISTER_PARAMETER(SimpleCropParam);
+DMLC_REGISTER_PARAMETER(SimpleClipParam);
```
Comment (piiswrong) [inline src/operator/tensor/matrix_op.cc]: No need for simple here

File: src/operator/tensor/matrix_op-inl.h [hunk 5]
```diff
@@ -704,6 +715,59 @@ inline bool CropShape(const nnvm::NodeAttrs& attrs,
...
+  using namespace mshadow::expr;
+  const SimpleClipParam& param = nnvm::get<SimpleClipParam>(attrs.parsed);
+  CHECK_EQ(inputs[0].type_flag_, outputs[0].type_flag_);
+  Stream<xpu> *s = ctx.get_stream<xpu>();
+
+  MSHADOW_TYPE_SWITCH(outputs[0].type_flag_, DType, {
+    Tensor<xpu, 1, DType> in = inputs[0].FlatTo1D<xpu, DType>(s);
+    Tensor<xpu, 1, DType> out = outputs[0].FlatTo1D<xpu, DType>(s);
+
+    out = F<mshadow_op::ClipMax::mshadow_op>(
```
Comment (piiswrong) [inline src/operator/tensor/matrix_op-inl.h]: don't use mshadow. Follow the batch_take example

File: src/operator/mshadow_op.h [hunk 6]
```diff
@@ -886,6 +886,32 @@ struct nanprod_grad {
   }
 };
 
+struct ClipMin {
```
Comment (piiswrong) [inline src/operator/mshadow_op.h]: use a mshadow op

File: src/operator/tensor/matrix_op.cc [hunk 7]
```diff
@@ -282,5 +283,24 @@ NNVM_REGISTER_OP(_backward_batch_dot)
 .set_attr<nnvm::TIsBackward>("TIsBackward", true)
 .set_attr<FCompute>("FCompute<cpu>", BatchDotBackward_<cpu>);
 
+NNVM_REGISTER_OP(clip)
+.MXNET_DESCRIBE("Clip ndarray elements to range (a_min, a_max)")
+.set_num_inputs(1)
+.set_num_outputs(1)
+.set_attr_parser(ParamParser<ClipParam>)
+.set_attr<nnvm::FInferShape>("FInferShape", ClipShape)
```
Comment (piiswrong) [inline src/operator/tensor/matrix_op.cc]: ElemwiseShape<1, 1>?
